Deprecation warning should print location in build file

huxi opened the pull request:
Any of the checked boxes below indicate that I took action:

- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).
- [x] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [x] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:

- [x] Before submitting a pull request, I started a discussion on the
[Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev)
or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [ ] I wrote a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
can be found in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
I ran a build on your local machine via the command
`./gradlew quickCheck <impacted-subproject>:check`.
- [ ] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

This PR has been discussed in [Way to find root cause of a Gradle deprecation warning is arcane knowledge](https://discuss.gradle.org/t/way-to-find-root-cause-of-a-gradle-deprecation-warning-is-arcane-knowledge/18838).

wolfs commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:
```diff
@@ -134,6 +134,7 @@ public static void propagateSystemProperty(String name) {
 
     protected boolean noExplicitTmpDir;
     protected boolean noExplicitNativeServicesDir;
+    protected boolean noFullDeprecationStackTrace;
```
Please call this field `fullDeprecationStackTrace` and initialize it to true. `eagerClassLoaderCreationChecksOn` is a better template for a flag like that. Double negation makes my head hurt.

wolfs commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/GradleExecuter.java:
```diff
@@ -180,6 +180,11 @@
     /**
+     * Don't activate full deprecation stack trace.
+     */
+    GradleExecuter withNoFullDeprecationStackTrace();
```
better: `withFullDeprecationStackTraceDisabled`

wolfs commented on subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java:
```diff
@@ -42,9 +42,6 @@ protected Boolean initialValue() {
     };
 
-    public static final String ORG_GRADLE_DEPRECATION_TRACE_PROPERTY_NAME = "org.gradle.deprecation.trace";
-    public static final String INCUBATION_MESSAGE = "%s is an incubating feature.";
```
Why did you remove the constant?

huxi replied:
`INCUBATION_MESSAGE`?
Because it was only used by the local `incubatingFeatureUsed(String, String)` and `ParallelForkingGradleHandle` in integration tests.

It makes me feel safer if tests don't use "insider knowledge" like that constant.

If this constant is changed then I would prefer to see failing tests if they somehow depend on the value rather than tests that are still green but should have been reviewed.

And I inlined the variable because all other similar methods have their pattern defined inline.

wolfs commented on subprojects/logging/src/integTest/groovy/org/gradle/DeprecationHandlingIntegrationTest.groovy:
```diff
@@ -45,9 +80,9 @@ class DeprecatedTask extends DefaultTask {
 
         then:
-        output.contains("Build file '$buildFile': line 3")
+        output.contains("Build file '$buildFile': line 10")
```
Actually, line 3 and line 6 should still be reported as the first calls to a deprecated method. Regarding the stacktrace it seems to me that the call to `nagUserOfDiscontinuedMethod` should never be reported in the message but what we want to report is the call to the method calling `nagUserOfDiscontinuedMethod`. This would actually mean to always drop the first (or last depending on your point of view) Stack frame and then reporting the next. This should happen while keeping the special handling for build files.

huxi replied:
This is only happening because of the way the integration test is structured.

The message is triggered inside of a `.gradle` file which would not be the case in a real life situation since the code calling the `nagUserOfDiscontinuedMethod` would be in the Gradle codebase instead.

If `someFeature()` and `otherFeature()` were defined in either a Java or Groovy file then the result would be what you expected. As it is, `LoggingDeprecatedFeatureHandler.isGradleScriptElement` is returning true for line 10, so line 3, one stack element later, is not printed.

You can see that everything is working correctly in the new integration tests using the realistic build file you provided.

wolfs commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecatedFeatureUsage.java:
```diff
@@ -48,41 +48,45 @@ public String getMessage() {
-                || stackTrace[caller].getClassName().startsWith("jdk.internal.")) {
-            caller++;
+    private static int skipSystemStackElements(StackTraceElement[] stackTrace, int caller) {
+        for (; caller < stackTrace.length; caller++) {
```
You don't like while, do you?

huxi replied:
I mainly dislike repeated identical code like `stackTrace[caller].getClassName()` and would always introduce a local variable in such a case.

So I needed the scope...

And since the body of the loop was incrementing `caller` anyway... :)

wolfs commented on subprojects/logging/src/test/groovy/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandlerTest.groovy:
```diff
@@ -16,42 +16,235 @@
+
+        and:
+        events[0].message == 'feature1'
+        events[1].message == 'feature2'
```
Check for the logLevel/category, too. The old assertion did this.

wolfs commented on subprojects/logging/src/test/groovy/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandlerTest.groovy:
```diff
@@ -16,42 +16,235 @@
+        message.startsWith(TextUtil.toPlatformLineSeparators('fake\n'))
+        message.contains('some.GradleScript.foo')
+
+        if (deprecationTraceProperty) {
```
Please write two tests for that. Having if like that in the test method is a test smell.

wolfs commented on subprojects/logging/src/test/groovy/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandlerTest.groovy:
```diff
@@ -16,42 +16,235 @@
-        def usage = new DeprecatedFeatureUsage("feature", LoggingDeprecatedFeatureHandlerTest)
+    @Unroll
+    def 'logs fake call Gradle script element first and #deprecationTracePropertyName=#deprecationTraceProperty'() {
+        System.setProperty(deprecationTracePropertyName, '' + deprecationTraceProperty)
```
Use `SetSystemProperties` as rule for setting system properties.

huxi replied:
Is any special handling necessary to make `SetSystemProperties` work?
Because the `@Rule` is already applied to the spec.

wolfs replied:
Oh sorry - did not see that. All good then.

wolfs commented:
Thank you very much for your PR. I added some inline comments - please address those. My main pain point is that currently the wrong location would be logged for build scripts.

huxi commented:
I answered that main point inline and will address the remainder later.

huxi commented on subprojects/logging/src/test/groovy/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandlerTest.groovy:
```diff
@@ -16,42 +16,262 @@
+        and:
+        def event = outputEventListener.events[0]
+        event.message == 'feature'
+        event.logLevel == LogLevel.WARN
```
logLevel is now checked in a separate test.

huxi commented:
I think I addressed all things you mentioned.
Let me know if I missed something.

huxi commented:
@wolfs I added `DeprecatedPlugin` and `DeprecatedTask` to `testFixtures` as well as tests using those classes in `DeprecationHandlingIntegrationTest`.

They are testing the same things as the already existing tests but without breaking during Gradle 4.0 upgrade and without being "unrealistic" in that the `DeprecationLogger` isn't called in the `.gradle` file anymore.

I'd suggest to replace the current tests with my new ones but wouldn't delete the existing tests without your OK, especially since you see issues with the way deprecation warnings originating from a Gradle script are currently handled.

I also added several `// TODO:` comments. Those regard a direct call of a task of an applied plugin. The task does not have a location in the script file since the plugin is simply applied, creating the task, and the task is executed via command line argument.

In that case, the output looks currently like this:

    :deprecated
    The deprecated task has been deprecated and is scheduled to be removed in Gradle 4.0. Please use the foobar task instead.
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

That's useless noise so I would suggest to not print the stack trace element at all if no script source can be resolved.

wolfs commented:
Regarding to have `-s` to cause the full stacktraces printed out: You can just pass `startParameter.getShowStacktrace()` to `DeprecationLogger` in `org.gradle.initialization.DefaultGradleLauncherFactory#doNewInstance`.
When this is implemented, should we leave the old method to trigger stacktraces (`-Dorg.gradle.deprecation.trace=true`) in place? What do you think?

wolfs commented:
Regarding the stacktraces: I think we should use something like `GroovyStackTraceSpec` to filter the stacktrace before outputting it. Then the example you pointed out above would not happen at all.

wolfs commented:
You can replace the old tests with the new ones.

huxi commented:
I just had to remove a piece of special handling code to properly filter the stack trace element.

I think we should keep `org.gradle.deprecation.trace` because it comes in quite handy in case of tests and won't hurt anybody.

I'd suggest to first check if it has been set at all in the system properties and if so, that it would override a probably existing `-s`. That check isn't performed at the moment so not setting the property results in `false` regardless of anything else.

You wouldn't be forced to get the full stack traces if you aren't actually interested in them.

wolfs commented:
Regarding the stack trace: There are two parts of it. I think you are saying that we do not want to output locations if there is no build script (*.gradle) in the stack trace. I think it still would be valuable to know where this deprecation is called. I want to output the location, but without the internal method (like `invoke0`) calls. This is where filtering the stack trace would be useful - and I think filtering the stacktrace is also the right thing to do when showing the full stack trace. I do not think that these internal calls are interesting for a plugin developer/build script user. This is why we filter those out for tests, too.

huxi commented:
I examined the stack trace of my test `DeprecatedPlugin and DeprecatedTask - without full stacktrace` for the `"deprecated"` task execution. It triggers the situation we are discussing because `"deprecated"` is a task introduced by `DeprecatedPlugin` so it does not have a source location in the script file(s).

There is really nothing in that whole stack trace (see below) that would be of any help to a normal user, unfortunately. And other users looking for something very specific should be fine with just using `-s/-S/-Dorg.gradle.deprecation.trace=true` in that case.

    The deprecated task has been deprecated and is scheduled to be removed in Gradle 4.0. Please use the foobar task instead.
        at org.gradle.internal.deprecated.DeprecatedTask.causeDeprecationWarning(DeprecatedTask.java:26)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:133)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:126)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:115)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:623)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:606)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:203)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:185)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:74)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:55)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:32)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:110)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
        at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:186)
        at org.gradle.internal.Factories$1.create(Factories.java:22)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:53)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:183)
        at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:33)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:112)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:63)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:106)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:92)
        at org.gradle.launcher.exec.GradleBuildController.run(GradleBuildController.java:66)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:41)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:85)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:52)

huxi commented:
I *think* this is ready for a merge now.

bmuschko commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:
```diff
@@ -705,7 +709,7 @@ private boolean noDaemonArgumentGiven() {
             properties.put(NativeServices.NATIVE_DIR_OVERRIDE, buildContext.getNativeServicesDir().getAbsolutePath());
         }
-        properties.put(DeprecationLogger.ORG_GRADLE_DEPRECATION_TRACE_PROPERTY_NAME, "true");
+        properties.put(LoggingDeprecatedFeatureHandler.ORG_GRADLE_DEPRECATION_TRACE_PROPERTY_NAME, ""+fullDeprecationStackTrace);
```
Should use `Boolean.toString(boolean)`.

huxi replied:
Done.

wolfs commented on subprojects/logging/logging.gradle:
```diff
@@ -14,10 +14,13 @@ dependencies {
+    testFixturesCompile project(":core")
+
+    testCompile project(":internalTesting")
+    testRuntime project(":jetty") // needed by DeprecationHandlingIntegrationTest
```
This is not needed anymore and should be removed.

huxi replied:
Excellent catch!

wolfs commented:
Hi @huxi,

this has been merged and will be included in Gradle 3.2. Thank you very much and keep up the good work!

wolfs commented:
Internal review: https://code-review.gradle.org/cru/REVIEW-6216
